spec.bs

```diff
@@ -5863,6 +5832,204 @@ an [=origin=] |topLevelOrigin|, and a [=string=] |slotSizeQueryParam|:
+  1. [=list/Extend=] |adComponentRenderURLs| with |entry|'s [=trusted scoring signals request/
+    ad component URLs=].
+1. Return the result of [=building trusted scoring signals url=] with |firstEntry|'s [=trusted
+  scoring signals request/base URL=], |renderURLs|, |adComponentRenderURLs|, |firstEntry|'s
+  [=trusted scoring signals request/seller experiment group id=],
+  |firstEntry|'s [=trusted scoring signals request/top level origin=].
+
+</div>
+
+<div algorithm>
+To <dfn>check if trusted scoring signals batch honors URL length limit</dfn> given a
+[=trusted scoring signals batcher=] |batcher| and a non-empty [=list=] of [=trusted
+scoring signals requests=] |entriesToBatch|:
+
+1. If the [=list/size=] of |entriesToBatch| is 1, return true.
+1. If |batcher|'s [=trusted scoring signals batcher/url length limit=] is 0, return true.
+1. Return whether the [=string/length=] of the result of [=building batched trusted scoring signals
+  url=] for |entriesToBatch| is &le; |batcher|'s [=trusted scoring signals batcher/url length
+  limit=].
+
+</div>
+
+<div algorithm>
+
+To <dfn>batch and fetch trusted scoring signals</dfn> given a [=trusted scoring signals batcher=]
+|batcher|:
+
+1. Until this object is no longer needed:
+  1. Wait until |batcher|'s [=trusted scoring signals batcher/request queue=] is no longer empty or
+    some heuristically chosen amount of time has passed.
+  1. Atomically transfer all entries in |batcher|'s [=trusted scoring signals batcher/
+    request queue=] to |incomingRequests|, leaving |batcher|'s [=trusted scoring signals batcher/
+    request queue=] empty.
+  1. For each |request| in |incomingRequests|:
+    1. Let |key| be (|request|'s [=trusted scoring signals request/seller script fetcher=],
+      |request|'s [=trusted scoring signals request/base url=], |request|'s [=trusted scoring
+      signals request/seller experiment group id=], |request|'s [=trusted scoring signals request/
+      top level origin=]).
+    1. If |batcher|'s [=trusted scoring signals batcher/request map=][|key|] does not [=map/exist=],
+      then [=map/set=] |batcher|'s [=trusted scoring signals batcher/request map=][|key|] to an
+      empty [=list=].
+    1. [=list/Append=] |request| to |batcher|'s [=trusted scoring signals batcher/request map=]
+      [|key|].
+  1. Some number of times, heuristically, select a |key| and a non-empty sublist of |batcher|'s
+    [=trusted scoring signals batcher/request map=] [|key|], called |entriesToBatch|, such that
+    [=checking if trusted scoring signals batch honors URL length limit=] on |batcher| and
+    |entriesToBatch| returns true:
+
+      Note: implementations are free to wait to collect more requests to merge or send them
+      individually, but need to take into account the configured URL length limit if they do combine
+      requests.
+
+      1. [=list/Remove=] |entriesToBatch| from |batcher|'s [=trusted scoring signals batcher/
+        request map=] [|key|]

```

Comment:
Suggest removing from the map if empty.  The "non-empty" sublist from the selection step implies that we can't select a key with an empty list, but still seems a bit weird to have empty lists floating around.

Reply:
Yeah, not doing this is an oversight, added.